### ISSUE TITLE
Delay pre-check short-circuit until all pre-cheks are done?

### ISSUE THREAD

**Issue by brendon** [palkan/action_policy#21]
Currently calling `allow!` or `deny!` in a pre-check short-circuits and returns the verdict immediately. Would you consider the concept of perhaps delaying the short-circuit until all pre-checks have run (especially useful in the case of some pre-checks being in sub-classes.

Here's my example:

ApplicationPolicy has a pre-check for `user.administrator?` and will `allow!` if so.
A SubClass has another pre-check for `object.deleted?` and will `deny!` if so. Unfortunately this never sees the light of day with an administrator and they'll be able to muck around with deleted objects.

Perhaps instead, all pre-checks can run, then if any returned `deny!` then deny, otherwise if any return `allow!` then allow. If none of the pre-checks return a verdict then process the policy action as usual.

What do you think?

**Comment by palkan** [comment]
IMO, short-curcuit-_ness_ is the main feature of pre-checks.

I think, a better solution could be to add an ability to _prepend_ checks, i.e.:

```ruby
class ClassPolicy < ApplicationPolicy
  pre_check :allow_admins
end

class SubClassPolicy < ClassPolicy
  pre_check :verify_deleted, prepend: true
end
```

What do you think?

**Comment by brendon** [comment]
That could get complicated quite quickly. Someone will ask to be able to insert one pre-check before/after another etc... and we'll end up with middleware! :D

What if we were able to group the pre-checks that `deny!` and `allow!` and execute the `deny`'s first, followed by the `allow`'s? It's a bit weird though.

I think the idea here is that `deny!` should take precedence over `allow!`, at least that's my thinking. Is that too big of an assumption?

**Comment by palkan** [comment]
> I think the idea here is that deny! should take precedence over allow!, at least that's my thinking. Is that too big of an assumption?

Don't think the they would it should work. Pre-checks are not meant specifically DRY your code, but to extract common _global_ rules.

The main use-case for them is admin/superadmin handling or stuff like that.

Overusing pre-checks makes code more complicated.

Probably, in your case it make sense to re-define the pre-check:

```ruby
class SubClassPolicy < ApplicationPolicy
  # remove previously defined admin check
  skip_pre_check :allow_admins?

  # add new check which also check for `object.deleted?`
  pre_check :allow_admins_unless_deleted
end
```

While I was writing this I thought about the idea of pre-check modifiers:

```ruby
class SubClassPolicy < ApplicationPolicy
  pre_check :allow_admins?, unless: -> { object.deleted? }
end
```

**Comment by brendon** [comment]
That last bit of code is an interesting idea :) In the end I realised that my existing code for determining `object.has_permission_to(:edit, user)` takes into account the users administrative status so I was able to skip that pre-check and just keep the deleted one. So I don't have a need for this anymore but perhaps it'll crop up for me again as I spread this across the rest of the application. If so, I'll put in a PR :) Feel free to close this if you think it's sorted.
